File: frontend/lib/core/models/academic.dart
class Semester {
  final int id;
  final String name;
  final String startDate;
  final String endDate;
  final bool isActive;
  final int universityId;
  final int term; // 1 = first semester of the year, 2 = second

  const Semester({
    required this.id, required this.name, required this.startDate,
    required this.endDate, required this.isActive, required this.universityId,
    this.term = 1,
  });

  factory Semester.fromJson(Map<String, dynamic> json) => Semester(
        id: json['id'] as int,
        name: json['name'] as String,
        startDate: json['start_date'] as String,
        endDate: json['end_date'] as String,
        isActive: json['is_active'] as bool,
        universityId: json['university_id'] as int,
        term: (json['term'] as int?) ?? 1,
      );
}

class TimeSlot {
  final int id;
  final String dayOfWeek;
  final String startTime;
  final String endTime;
  final int semesterId;

  const TimeSlot({
    required this.id, required this.dayOfWeek, required this.startTime,
    required this.endTime, required this.semesterId,
  });

  factory TimeSlot.fromJson(Map<String, dynamic> json) => TimeSlot(
        id: json['id'] as int,
        dayOfWeek: json['day_of_week'] as String,
        startTime: json['start_time'] as String,
        endTime: json['end_time'] as String,
        semesterId: json['semester_id'] as int,
      );

  String get label => '$startTime – $endTime';
}

class Level {
  final int id;
  final int number;
  final int departmentId;

  const Level({required this.id, required this.number, required this.departmentId});

  factory Level.fromJson(Map<String, dynamic> json) => Level(
        id: json['id'] as int,
        number: json['number'] as int,
        departmentId: json['department_id'] as int,
      );

  String get label => 'Level $number';
}

class StudyClass {
  final int id;
  final String name;
  final int population;
  final int levelId;

  const StudyClass({
    required this.id, required this.name,
    required this.population, required this.levelId,
  });

  factory StudyClass.fromJson(Map<String, dynamic> json) => StudyClass(
        id: json['id'] as int,
        name: json['name'] as String,
        population: json['population'] as int,
        levelId: json['level_id'] as int,
      );
}

File: backend/app/schemas/academic.py
from typing import Optional
from pydantic import BaseModel


class SemesterCreate(BaseModel):
    name: str
    start_date: str
    end_date: str
    university_id: int
    term: int = 1  # 1 = first semester of the year, 2 = second


class SemesterUpdate(BaseModel):
    name: Optional[str] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    is_active: Optional[bool] = None
    term: Optional[int] = None


class SemesterOut(BaseModel):
    id: int
    name: str
    start_date: str
    end_date: str
    is_active: bool
    university_id: int
    term: int

    model_config = {"from_attributes": True}


class TimeSlotCreate(BaseModel):
    day_of_week: str
    start_time: str
    end_time: str
    semester_id: int


class TimeSlotOut(BaseModel):
    id: int
    day_of_week: str
    start_time: str
    end_time: str
    semester_id: int

    model_config = {"from_attributes": True}


class LevelCreate(BaseModel):
    number: int
    department_id: int


class LevelOut(BaseModel):
    id: int
    number: int
    department_id: int

    model_config = {"from_attributes": True}


class ClassCreate(BaseModel):
    name: str
    population: int
    level_id: int


class ClassUpdate(BaseModel):
    name: Optional[str] = None
    population: Optional[int] = None


class ClassOut(BaseModel):
    id: int
    name: str
    population: int
    level_id: int

    model_config = {"from_attributes": True}


class ClassGroupCreate(BaseModel):
    name: str
    size: int
    class_id: int


class ClassGroupUpdate(BaseModel):
    name: Optional[str] = None
    size: Optional[int] = None


class ClassGroupOut(BaseModel):
    id: int
    name: str
    size: int
    class_id: int

    model_config = {"from_attributes": True}
